**Problem Statement- we have to predict number of positive & negative reviews based on sentiments by using different classification models**

In [ ]:
from nltk.tokenize import word_tokenize
from nltk.tokenize import sent_tokenize
from nltk.corpus import stopwords
from nltk import WordNetLemmatizer
from nltk import PorterStemmer
import nltk
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [ ]:
data=pd.read_csv('Dataset.csv')
data

review sentiment
0     One of the other reviewers has mentioned that ...  positive
1     A wonderful little production. <br /><br />The...  positive
2     I thought this was a wonderful way to spend ti...  positive
3     Basically there's a family where a little boy ...  negative
4     Petter Mattei's "Love in the Time of Money" is...  positive
...                                                 ...       ...
9995  Fun, entertaining movie about WWII German spy ...  positive
9996  Give me a break. How can anyone say that this ...  negative
9997  This movie is a bad movie. But after watching ...  negative
9998  This is a movie that was probably made to ente...  negative
9999  Smashing film about film-making. Shows the int...  positive

[10000 rows x 2 columns]

In [ ]:
data['sentiment'].value_counts()

sentiment
positive    5028
negative    4972
Name: count, dtype: int64

In [ ]:
data.shape

(10000, 2)

In [ ]:
data.describe()

review sentiment
count                                               10000     10000
unique                                               9983         2
top     Beautiful film, pure Cassavetes style. Gena Ro...  positive
freq                                                    2      5028

In [ ]:
data.info

<bound method DataFrame.info of                                                  review sentiment
0     One of the other reviewers has mentioned that ...  positive
1     A wonderful little production. <br /><br />The...  positive
2     I thought this was a wonderful way to spend ti...  positive
3     Basically there's a family where a little boy ...  negative
4     Petter Mattei's "Love in the Time of Money" is...  positive
...                                                 ...       ...
9995  Fun, entertaining movie about WWII German spy ...  positive
9996  Give me a break. How can anyone say that this ...  negative
9997  This movie is a bad movie. But after watching ...  negative
9998  This is a movie that was probably made to ente...  negative
9999  Smashing film about film-making. Shows the int...  positive

[10000 rows x 2 columns]>

In [ ]:
data.isnull().sum()

review       0
sentiment    0
dtype: int64

Removing Punctuation

In [ ]:
from string import punctuation

In [ ]:
def pun(x):
    return ''.join([i for i in x if i not in punctuation])

In [ ]:
data['Punct']=data['review'].apply(lambda x:pun(x))


Removing Emoji

In [ ]:
!pip install emoji

In [ ]:
import emoji

In [ ]:
data['Emoji']=data['Punct'].apply(lambda x:emoji.replace_emoji(x,''))

In [ ]:
data

review sentiment  \
0     One of the other reviewers has mentioned that ...  positive   
1     A wonderful little production. <br /><br />The...  positive   
2     I thought this was a wonderful way to spend ti...  positive   
3     Basically there's a family where a little boy ...  negative   
4     Petter Mattei's "Love in the Time of Money" is...  positive   
...                                                 ...       ...   
9995  Fun, entertaining movie about WWII German spy ...  positive   
9996  Give me a break. How can anyone say that this ...  negative   
9997  This movie is a bad movie. But after watching ...  negative   
9998  This is a movie that was probably made to ente...  negative   
9999  Smashing film about film-making. Shows the int...  positive   

                                                  Punct  \
0     One of the other reviewers has mentioned that ...   
1     A wonderful little production br br The filmin...   
2     I thought this was a wonderful way to spend ti...   
3     Basically theres a family where a little boy J...   
4     Petter Matteis Love in the Time of Money is a ...   
...                                                 ...   
9995  Fun entertaining movie about WWII German spy J...   
9996  Give me a break How can anyone say that this i...   
9997  This movie is a bad movie But after watching a...   
9998  This is a movie that was probably made to ente...   
9999  Smashing film about filmmaking Shows the inten...   

                                                  Emoji  
0     One of the other reviewers has mentioned that ...  
1     A wonderful little production br br The filmin...  
2     I thought this was a wonderful way to spend ti...  
3     Basically theres a family where a little boy J...  
4     Petter Matteis Love in the Time of Money is a ...  
...                                                 ...  
9995  Fun entertaining movie about WWII German spy J...  
9996  Give me a break How can anyone say that this i...  
9997  This movie is a bad movie But after watching a...  
9998  This is a movie that was probably made to ente...  
9999  Smashing film about filmmaking Shows the inten...  

[10000 rows x 4 columns]

Tokenization, Lemmitization, Stemming

In [ ]:
nltk.download('stopwords')
nltk.download('punkt')
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\LEGION\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\LEGION\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\LEGION\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [ ]:
LE=WordNetLemmatizer()

In [ ]:
sw=stopwords.words('english')

In [ ]:
Ps=PorterStemmer()

In [ ]:
def preprocess(x):
  a=x.lower()
  b=''.join([i for i in a if i not in punctuation])
  c=word_tokenize(b)
  d=' '.join([Ps.stem(i) for i in c if i not in stopwords.words('english')])
  return d

In [ ]:
data['Preprocess']=data['Emoji'].apply(lambda x:preprocess(x))
data

review sentiment  \
0     One of the other reviewers has mentioned that ...  positive   
1     A wonderful little production. <br /><br />The...  positive   
2     I thought this was a wonderful way to spend ti...  positive   
3     Basically there's a family where a little boy ...  negative   
4     Petter Mattei's "Love in the Time of Money" is...  positive   
...                                                 ...       ...   
9995  Fun, entertaining movie about WWII German spy ...  positive   
9996  Give me a break. How can anyone say that this ...  negative   
9997  This movie is a bad movie. But after watching ...  negative   
9998  This is a movie that was probably made to ente...  negative   
9999  Smashing film about film-making. Shows the int...  positive   

                                                  Punct  \
0     One of the other reviewers has mentioned that ...   
1     A wonderful little production br br The filmin...   
2     I thought this was a wonderful way to spend ti...   
3     Basically theres a family where a little boy J...   
4     Petter Matteis Love in the Time of Money is a ...   
...                                                 ...   
9995  Fun entertaining movie about WWII German spy J...   
9996  Give me a break How can anyone say that this i...   
9997  This movie is a bad movie But after watching a...   
9998  This is a movie that was probably made to ente...   
9999  Smashing film about filmmaking Shows the inten...   

                                                  Emoji  \
0     One of the other reviewers has mentioned that ...   
1     A wonderful little production br br The filmin...   
2     I thought this was a wonderful way to spend ti...   
3     Basically theres a family where a little boy J...   
4     Petter Matteis Love in the Time of Money is a ...   
...                                                 ...   
9995  Fun entertaining movie about WWII German spy J...   
9996  Give me a break How can anyone say that this i...   
9997  This movie is a bad movie But after watching a...   
9998  This is a movie that was probably made to ente...   
9999  Smashing film about filmmaking Shows the inten...   

                                             Preprocess  
0     one review mention watch 1 oz episod youll hoo...  
1     wonder littl product br br film techniqu unass...  
2     thought wonder way spend time hot summer weeke...  
3     basic there famili littl boy jake think there ...  
4     petter mattei love time money visual stun film...  
...                                                 ...  
9995  fun entertain movi wwii german spi juli andrew...  
9996  give break anyon say good hockey movi know mov...  
9997  movi bad movi watch endless seri bad horror mo...  
9998  movi probabl made entertain middl school earli...  
9999  smash film filmmak show intens strang relation...  

[10000 rows x 5 columns]

Removing the html strips

In [ ]:
!pip install beautifulsoup4

In [ ]:
from bs4 import BeautifulSoup

In [ ]:
#Removing the html strips
def strip_html(text):
    soup = BeautifulSoup(text, "html.parser")
    return soup.get_text()

#Removing the noisy text
def denoise_text(text):
    text = strip_html(text)
    return text

In [ ]:
#Apply function on review column
data['Final']=data['Preprocess'].apply(denoise_text)
data

review sentiment  \
0     One of the other reviewers has mentioned that ...  positive   
1     A wonderful little production. <br /><br />The...  positive   
2     I thought this was a wonderful way to spend ti...  positive   
3     Basically there's a family where a little boy ...  negative   
4     Petter Mattei's "Love in the Time of Money" is...  positive   
...                                                 ...       ...   
9995  Fun, entertaining movie about WWII German spy ...  positive   
9996  Give me a break. How can anyone say that this ...  negative   
9997  This movie is a bad movie. But after watching ...  negative   
9998  This is a movie that was probably made to ente...  negative   
9999  Smashing film about film-making. Shows the int...  positive   

                                                  Punct  \
0     One of the other reviewers has mentioned that ...   
1     A wonderful little production br br The filmin...   
2     I thought this was a wonderful way to spend ti...   
3     Basically theres a family where a little boy J...   
4     Petter Matteis Love in the Time of Money is a ...   
...                                                 ...   
9995  Fun entertaining movie about WWII German spy J...   
9996  Give me a break How can anyone say that this i...   
9997  This movie is a bad movie But after watching a...   
9998  This is a movie that was probably made to ente...   
9999  Smashing film about filmmaking Shows the inten...   

                                                  Emoji  \
0     One of the other reviewers has mentioned that ...   
1     A wonderful little production br br The filmin...   
2     I thought this was a wonderful way to spend ti...   
3     Basically theres a family where a little boy J...   
4     Petter Matteis Love in the Time of Money is a ...   
...                                                 ...   
9995  Fun entertaining movie about WWII German spy J...   
9996  Give me a break How can anyone say that this i...   
9997  This movie is a bad movie But after watching a...   
9998  This is a movie that was probably made to ente...   
9999  Smashing film about filmmaking Shows the inten...   

                                             Preprocess  \
0     one review mention watch 1 oz episod youll hoo...   
1     wonder littl product br br film techniqu unass...   
2     thought wonder way spend time hot summer weeke...   
3     basic there famili littl boy jake think there ...   
4     petter mattei love time money visual stun film...   
...                                                 ...   
9995  fun entertain movi wwii german spi juli andrew...   
9996  give break anyon say good hockey movi know mov...   
9997  movi bad movi watch endless seri bad horror mo...   
9998  movi probabl made entertain middl school earli...   
9999  smash film filmmak show intens strang relation...   

                                                  Final  
0     one review mention watch 1 oz episod youll hoo...  
1     wonder littl product br br film techniqu unass...  
2     thought wonder way spend time hot summer weeke...  
3     basic there famili littl boy jake think there ...  
4     petter mattei love time money visual stun film...  
...                                                 ...  
9995  fun entertain movi wwii german spi juli andrew...  
9996  give break anyon say good hockey movi know mov...  
9997  movi bad movi watch endless seri bad horror mo...  
9998  movi probabl made entertain middl school earli...  
9999  smash film filmmak show intens strang relation...  

[10000 rows x 6 columns]

Dropping Unnessary Columns

In [ ]:
data.drop(['review', 'sentiment','Punct','Emoji','Preprocess'], axis=1,inplace=True)
data

Final
0     one review mention watch 1 oz episod youll hoo...
1     wonder littl product br br film techniqu unass...
2     thought wonder way spend time hot summer weeke...
3     basic there famili littl boy jake think there ...
4     petter mattei love time money visual stun film...
...                                                 ...
9995  fun entertain movi wwii german spi juli andrew...
9996  give break anyon say good hockey movi know mov...
9997  movi bad movi watch endless seri bad horror mo...
9998  movi probabl made entertain middl school earli...
9999  smash film filmmak show intens strang relation...

[10000 rows x 1 columns]

Vectorizer

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [ ]:
Tif=TfidfVectorizer()

In [ ]:
new=Tif.fit_transform(data['Final']).toarray()

In [ ]:
NewData=pd.DataFrame(new,columns=Tif.get_feature_names_out())
NewData.head(3)

00  000  00000110  0001  0069  007   01  010  01000  0126  ...  étcother  \
0  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...       0.0   
1  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...       0.0   
2  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...       0.0   

   êtrepeutêtr   ís  ísnt   ïn  óli  önsjön  über  überwoman  ünfaith  
0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0      0.0  
1          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0      0.0  
2          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0      0.0  

[3 rows x 55285 columns]

In [ ]:
TargetData=pd.read_csv('Dataset.csv')
TargetData

review sentiment
0     One of the other reviewers has mentioned that ...  positive
1     A wonderful little production. <br /><br />The...  positive
2     I thought this was a wonderful way to spend ti...  positive
3     Basically there's a family where a little boy ...  negative
4     Petter Mattei's "Love in the Time of Money" is...  positive
...                                                 ...       ...
9995  Fun, entertaining movie about WWII German spy ...  positive
9996  Give me a break. How can anyone say that this ...  negative
9997  This movie is a bad movie. But after watching ...  negative
9998  This is a movie that was probably made to ente...  negative
9999  Smashing film about film-making. Shows the int...  positive

[10000 rows x 2 columns]

In [ ]:
TargetData = TargetData.drop('review', axis=1)
TargetData

sentiment
0     positive
1     positive
2     positive
3     negative
4     positive
...        ...
9995  positive
9996  negative
9997  negative
9998  negative
9999  positive

[10000 rows x 1 columns]

Label Encoding

In [ ]:
from sklearn.preprocessing import LabelEncoder

In [ ]:
label_encoder = LabelEncoder()

In [ ]:
TargetData['encoded_sentiments'] = label_encoder.fit_transform(TargetData['sentiment'])


In [ ]:
TargetData

sentiment  encoded_sentiments
0     positive                   1
1     positive                   1
2     positive                   1
3     negative                   0
4     positive                   1
...        ...                 ...
9995  positive                   1
9996  negative                   0
9997  negative                   0
9998  negative                   0
9999  positive                   1

[10000 rows x 2 columns]

In [ ]:
TargetData = TargetData.drop('sentiment', axis=1)
TargetData

encoded_sentiments
0                      1
1                      1
2                      1
3                      0
4                      1
...                  ...
9995                   1
9996                   0
9997                   0
9998                   0
9999                   1

[10000 rows x 1 columns]

Final Dataset

In [ ]:
FinalData=pd.concat((NewData,TargetData),axis=1)
FinalData.head()

00  000  00000110  0001  0069  007   01  010  01000  0126  ...  \
0  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
1  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
2  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
3  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
4  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   

   êtrepeutêtr   ís  ísnt   ïn  óli  önsjön  über  überwoman  ünfaith  \
0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0      0.0   
1          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0      0.0   
2          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0      0.0   
3          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0      0.0   
4          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0      0.0   

   encoded_sentiments  
0                   1  
1                   1  
2                   1  
3                   0  
4                   1  

[5 rows x 55286 columns]

Dividing Data into X & y

In [ ]:
X=FinalData.drop('encoded_sentiments',axis=1)
y=FinalData['encoded_sentiments']

In [ ]:
X

00  000  00000110  0001  0069  007   01  010  01000  0126  ...  \
0     0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
1     0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
2     0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
3     0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
4     0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
...   ...  ...       ...   ...   ...  ...  ...  ...    ...   ...  ...   
9995  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
9996  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
9997  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
9998  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   
9999  0.0  0.0       0.0   0.0   0.0  0.0  0.0  0.0    0.0   0.0  ...   

      étcother  êtrepeutêtr   ís  ísnt   ïn  óli  önsjön  über  überwoman  \
0          0.0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0   
1          0.0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0   
2          0.0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0   
3          0.0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0   
4          0.0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0   
...        ...          ...  ...   ...  ...  ...     ...   ...        ...   
9995       0.0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0   
9996       0.0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0   
9997       0.0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0   
9998       0.0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0   
9999       0.0          0.0  0.0   0.0  0.0  0.0     0.0   0.0        0.0   

      ünfaith  
0         0.0  
1         0.0  
2         0.0  
3         0.0  
4         0.0  
...       ...  
9995      0.0  
9996      0.0  
9997      0.0  
9998      0.0  
9999      0.0  

[10000 rows x 55285 columns]

In [ ]:
y

0       1
1       1
2       1
3       0
4       1
       ..
9995    1
9996    0
9997    0
9998    0
9999    1
Name: encoded_sentiments, Length: 10000, dtype: int32

Dividing Data into train_test

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(X,y,test_size=0.20,random_state=0)

Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression

#Create a logistic regression model
model = LogisticRegression()

#Train the model on the training set
model.fit(x_train, y_train)

#Make predictions on the test set
y_pred = model.predict(x_test)

#Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)

print(f"Accuracy: {accuracy}")
print("Classification Report:\n", report)
print("Confusion Matrix:\n", conf_matrix)

Accuracy: 0.857
Classification Report:
               precision    recall  f1-score   support

           0       0.88      0.83      0.85       979
           1       0.84      0.89      0.86      1021

    accuracy                           0.86      2000
   macro avg       0.86      0.86      0.86      2000
weighted avg       0.86      0.86      0.86      2000

Confusion Matrix:
 [[808 171]
 [115 906]]


Random Forest Classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier

#Create a Random Forest classifier
RFC = RandomForestClassifier(n_estimators=100, random_state=42)

#Train the model on the training set
RFC.fit(x_train, y_train)

#Make predictions on the test set
y_predd = RFC.predict(x_test)

#Evaluate the model
accuracy = accuracy_score(y_test, y_predd)
report = classification_report(y_test, y_predd)
conf_matrix = confusion_matrix(y_test, y_predd)

print(f"Accuracy: {accuracy}")
print("Classification Report:\n", report)
print("Confusion Matrix:\n", conf_matrix)

Accuracy: 0.8395
Classification Report:
               precision    recall  f1-score   support

           0       0.84      0.83      0.84       979
           1       0.84      0.85      0.84      1021

    accuracy                           0.84      2000
   macro avg       0.84      0.84      0.84      2000
weighted avg       0.84      0.84      0.84      2000

Confusion Matrix:
 [[815 164]
 [157 864]]


Logistic Regression gave accuarcy of 85% and Random Forest Classifier gave accuracy of 83%